### PR TITLE
Consuming 2.0.0 preview8

### PR DESCRIPTION
**Pull request by ptankTwilio**
Consuming TwilioVideo 2.0.0-preview8 -
1. Callkit sample app uses new `TVIDefaultAudioDevice`
2. Basic Quickstart logs error when subscription error occurs.
3. Deprecated 8.1 support.

### CODE REVIEW

File: Podfile [hunk 1]
```diff
@@ -16,7 +16,7 @@ abstract_target 'TwilioVideo' do
   end
 
   target 'VideoQuickStart' do
-    platform :ios, '8.1'
+    platform :ios, '9.0'
```
Comment (paynerc) [inline Podfile]: 🎉

File: Podfile [hunk 2]
```diff
@@ -26,12 +26,12 @@ abstract_target 'TwilioVideo' do
   end
 
   target 'ScreenCapturerExample' do
-    platform :ios, '8.1'
+    platform :ios, '9.0'
```
Comment (paynerc) [inline Podfile]: 🎉

File: Podfile [hunk 3]
```diff
@@ -26,12 +26,12 @@ abstract_target 'TwilioVideo' do
...
     project 'ScreenCapturerExample.xcproject'
   end
 
   target 'DataTrackExample' do
-    platform :ios, '8.1'
+    platform :ios, '9.0'
```
Comment (paynerc) [inline Podfile]: 🎉

File: VideoCallKitQuickStart/ViewController+CallKit.swift [hunk 4]
```diff
@@ -16,7 +16,9 @@ extension ViewController : CXProviderDelegate {
     func providerDidReset(_ provider: CXProvider) {
         logMessage(messageText: "providerDidReset:")
 
-        TVIAudioController.shared().stopAudio()
+        // AudioDevice is enabled by default
+        self.audioDevice.isEnabled = true
```
Comment (ceaglest) [inline VideoCallKitQuickStart/ViewController+CallKit.swift]: Don't we want it disabled by default? We should encourage users to be really sure that no audio can possibly play before CallKit has done its magic.

File: VideoCallKitQuickStart/ViewController+CallKit.swift [hunk 5]
```diff
@@ -191,6 +204,9 @@ extension ViewController {
             // Use the local media that we prepared earlier.
             builder.audioTracks = self.localAudioTrack != nil ? [self.localAudioTrack!] : [TVILocalAudioTrack]()
             builder.videoTracks = self.localVideoTrack != nil ? [self.localVideoTrack!] : [TVILocalVideoTrack]()
+
+            // Use the audio device
+            builder.audioDevice = self.audioDevice
```
Comment (ceaglest) [inline VideoCallKitQuickStart/ViewController+CallKit.swift]: Should we mention any more about this API - like there can only be a single audio device used?

File: VideoQuickStart/ViewController.swift [hunk 6]
```diff
@@ -426,6 +426,14 @@ extension ViewController : TVIRemoteParticipantDelegate {
...
         logMessage(messageText: "Participant \(participant.identity) disabled \(publication.trackName) audio track")
     }
+
+    func failedToSubscribe(toAudioTrack publication: TVIRemoteAudioTrackPublication, error: Error, for participant: TVIRemoteParticipant) {
+        logMessage(messageText: "FailedToSubscribe \(publication.trackName) audio track, error = \(String(describing: error))")
+    }
+
+    func failedToSubscribe(toVideoTrack publication: TVIRemoteVideoTrackPublication, error: Error, for participant: TVIRemoteParticipant) {
+        logMessage(messageText: "FailedToSubscribe \(publication.trackName) video track, error = \(String(describing: error))")
+    }
```
Comment (ceaglest) [inline VideoQuickStart/ViewController.swift]: Can you implement these callbacks in the other examples in the repo?

File: VideoCallKitQuickStart/ViewController.swift [hunk 7]
```diff
@@ -23,6 +23,7 @@ class ViewController: UIViewController {
     
     // Video SDK components
     var room: TVIRoom?
+    var audioDevice: TVIDefaultAudioDevice = TVIDefaultAudioDevice()
```
Comment (ceaglest) [inline VideoCallKitQuickStart/ViewController.swift]: This actually might be a good spot to mention about single audio devices.

File: VideoCallKitQuickStart/ViewController+CallKit.swift [hunk 8]
```diff
@@ -66,7 +73,12 @@ extension ViewController : CXProviderDelegate {
          * Configure the audio session, but do not start call audio here, since it must be done once
          * the audio session has been activated by the system after having its priority elevated.
          */
-        TVIAudioController.shared().configureAudioSession(.videoChatSpeaker)
+
+        // Stop the audio unit by setting isEnabled to `false`.
+        self.audioDevice.isEnabled = false;
+
+        // Configure the AVAudioSession by executign the audio device's `block`.
+        self.audioDevice.block()
```
Comment (ceaglest) [inline VideoCallKitQuickStart/ViewController+CallKit.swift]: Minor, you could move lines 72 to 81 into `performRoomConnect` since Starting and Answering a Call are the same W.R.T. audio.

File: VideoCallKitQuickStart/ViewController+CallKit.swift [hunk 9]
```diff
@@ -80,7 +92,8 @@ extension ViewController : CXProviderDelegate {
     func provider(_ provider: CXProvider, perform action: CXEndCallAction) {
         NSLog("provider:performEndCallAction:")
 
-        TVIAudioController.shared().stopAudio()
+        // AudioDevice is enabled by default
+        self.audioDevice.isEnabled = true;
```
Comment (ceaglest) [inline VideoCallKitQuickStart/ViewController+CallKit.swift]: I think we want it disabled by default.

File: VideoCallKitQuickStart/ViewController+CallKit.swift (hunk 4 above)
Comment (ptankTwilio) [reply VideoCallKitQuickStart/ViewController+CallKit.swift]: We can. I was thinking we will change it back to the default value after callkit use is done.

File: VideoCallKitQuickStart/ViewController+CallKit.swift (hunk 8 above)
Comment (ptankTwilio) [reply VideoCallKitQuickStart/ViewController+CallKit.swift]: I would keep outside to show it explicitly on how to configure audio device. `performRoomConnect` has the _general sdk code_ to connect to a Room.

File: VideoCallKitQuickStart/ViewController+CallKit.swift (hunk 9 above)
Comment (ptankTwilio) [reply VideoCallKitQuickStart/ViewController+CallKit.swift]: We can. I was thinking we will change it back to the default value after callkit use is done. WDYT? I thought, during the review, @idelgado was suggesting it to reset to TVIDefaultAudioDevice's property `enabled`'s default value to  `true`.

I can go either way on this.

File: VideoCallKitQuickStart/ViewController+CallKit.swift (hunk 5 above)
Comment (ptankTwilio) [reply VideoCallKitQuickStart/ViewController+CallKit.swift]: good idea 👍

File: VideoCallKitQuickStart/ViewController+CallKit.swift [hunk 10]
```diff
@@ -191,6 +204,14 @@ extension ViewController {
             // Use the local media that we prepared earlier.
             builder.audioTracks = self.localAudioTrack != nil ? [self.localAudioTrack!] : [TVILocalAudioTrack]()
             builder.videoTracks = self.localVideoTrack != nil ? [self.localVideoTrack!] : [TVILocalVideoTrack]()
+
+            /**
+             * Use the audio device. Please note that the SDK does not support the use of multiple audio devices at the
+             * same time. If you've already connected to a Room, then all future connection attempts must use the same
+             * TVIDefaultAudioDevice as the first Room. Once all the existing Rooms are disconnected you are free to
+             * choose a new audio device for your next connection attempt.
+             */
```
Comment (ptankTwilio) [inline VideoCallKitQuickStart/ViewController+CallKit.swift]: @ceaglest do you think mentioning this comment here as well is ok?  We are already mentioning this in ViewController.swift as well

File: VideoCallKitQuickStart/ViewController+CallKit.swift (hunk 8 above)
Comment (ceaglest) [reply VideoCallKitQuickStart/ViewController+CallKit.swift]: Sounds fine to me, just something I noticed reading the code. I think you are correct that being explicit about the configuration steps is best (even if some duplication is done).

File: VideoCallKitQuickStart/ViewController+CallKit.swift (hunk 10 above)
Comment (ceaglest) [reply VideoCallKitQuickStart/ViewController+CallKit.swift]: I think saving the main comment for the property declaration is fine. I would keep that one, and shorten this to:

`// Use the audio device that we created earlier. All connection attempts will use the same device.`

Or maybe even from the 2nd sentence from that.

File: VideoCallKitQuickStart/ViewController.swift [hunk 11]
```diff
@@ -23,6 +23,13 @@ class ViewController: UIViewController {
     
     // Video SDK components
     var room: TVIRoom?
+    /**
+     * We will create an audio device and use it in Room. Please note that the SDK does not support the use of multiple
+     * audio devices at the same time. If you've already connected to a Room, then all future connection attempts must
+     * use the same TVIDefaultAudioDevice as the first Room. Once all the existing Rooms are disconnected you are free to
+     * choose a new audio device for your next connection attempt.
+     */
```
Comment (ceaglest) [inline VideoCallKitQuickStart/ViewController.swift]: Looks good. The first sentence doesn't quite make sense grammar wise. Perhaps:

`We will create an audio device and manage it's lifecycle in response to CallKit events.`

File: VideoCallKitQuickStart/ViewController+CallKit.swift (hunk 4 above)
Comment (ceaglest) [reply VideoCallKitQuickStart/ViewController+CallKit.swift]: I guess I don't feel that strongly about it. Lets keep what you have, its consistent with the voice QuickStarts.
